tests/integration/test_dynamodb.py

```diff
@@ -258,6 +258,84 @@ def test_multiple_update_expressions(self, dynamodb):
             )
         assert ctx.match("ValidationException")
 
+    def test_invalid_query_index(self, dynamodb):
+        """Raises an exception when a query requests ALL_ATTRIBUTES,
+        but the index does not have a ProjectionType of ALL"""
+        table_name = f"test-table-{short_uid()}"
+        table = dynamodb.create_table(
+            TableName=table_name,
+            KeySchema=[{"AttributeName": "id", "KeyType": "HASH"}],
+            AttributeDefinitions=[
+                {"AttributeName": "id", "AttributeType": "S"},
+                {"AttributeName": "field_a", "AttributeType": "S"},
+            ],
+            ProvisionedThroughput={"ReadCapacityUnits": 5, "WriteCapacityUnits": 5},
+            Tags=TEST_DDB_TAGS,
+            GlobalSecondaryIndexes=[
+                {
+                    "IndexName": "field_a_index",
+                    "KeySchema": [{"AttributeName": "field_a", "KeyType": "HASH"}],
+                    "Projection": {"ProjectionType": "KEYS_ONLY"},
+                    "ProvisionedThroughput": {
+                        "ReadCapacityUnits": 1,
+                        "WriteCapacityUnits": 1,
+                    },
+                },
+            ],
+        )
+
+        with pytest.raises(Exception) as ctx:
+            table.query(
+                TableName=table_name,
+                IndexName="field_a_index",
+                KeyConditionExpression=Key("field_a").eq("xyz"),
+                Select="ALL_ATTRIBUTES",
+            )
+        assert ctx.match("ValidationException")
```

Comment:
nit: Let's better clean up and delete the table here at the end of the test method. (same in line 337 below)